types/jsreport-core/index.d.ts

```diff
@@ -15,20 +18,36 @@ declare namespace JsReport {
     type Recipe = "html";
 
     interface Template {
-        content: string;
+        content?: string;
         engine: Engine | string;
-        helpers: Helpers;
+        helpers?: Helpers;
         recipe: Recipe | string;
+        pathToEngine?: string;
     }
 
-    interface Options {
+    interface RequestOptions {
         preview?: boolean;
+        timeout?: number;
+        reportName?: string;
     }
 
     interface Request {
+        value?: boolean;
+        writable?: boolean;
+        configurable?: boolean;
+        enumerable?: boolean;
         template: Partial<Template>;
```

Comment:
If you don't want remove `?` at line 21,23,25, remove `Partial` here.
```ts
template: Template;
```

Reply:
changed it back to Partial since that appears to be needed to support downstream plugins

Reply:
Do downstream plugins allow you to do `jscore.request({ template: {}, ... })`, or is the `Partial` provided so that downstream plugins that augment `Template` don't make their properties required? With these plugins, are `content`, `recipe`, and `engine` still required?

If so you'd want a type more like this:

```suggestion
        template: Pick<Template, "content" | "recipe" | "engine">
            & Partial<Omit<Template, "content" | "recipe" | "engine">>;
```

That would enforce `content`, `recipe`, and `engine` to be required but ensure all other options are optional.

---

There are other ways of handling extensibility as well, such as "open-ended unions":

```ts
interface Template {
  ...
}
interface NamedTemplate {
  name: string;
}
interface TemplateRegistry {
  Template: Template;
  NamedTemplate: NamedTemplate;
}
type TemplateLike = TemplateRegistry[keyof TemplateRegistry]; // Template | NamedTemplate
```

Then, downstream plugins could define their own "template" extensions with their own decisions on whether some properties are required or optional:

```ts
// some downstream plugin
import { Template } from "jsreport-core";

// declare my plugin's template
export interface MyTemplate extends Template {
}

// add it to the registry
declare module "jsreport-core" {
  interface TemplateRegistry {
    MyTemplate: MyTemplate; // this adds it to the `TemplateLike` union, above.
  }
}
```

We use this "open-ended union" functionality in the TypeScript lib.es5.d.ts to define the `ArrayBufferLike` type which contains only `ArrayBuffer` when targeting ES5, but also includes `SharedArrayBuffer` when targeting ES2017 or later.

Reply:
I have been looking at implementing the TemplateRegistery which I think makes a lot of sense but it's going to break a lot downstream.